Use runtime.Copier instead of hardcoding api.Scheme

smarterclayton opened the pull request:
Allow other schemes to be supported by etcd_helper.go

runtime.Scheme.Copy() should be using the built in DeepCopy()

deads2k commented on pkg/conversion/cloner.go:
```diff
@@ -134,7 +134,10 @@ func (c *Cloner) deepCopy(src reflect.Value) (reflect.Value, error) {
 		return c.customDeepCopy(src, fv)
 	}
-	return c.defaultDeepCopy(src)
+	glog.Infof("> default deep copy of %v", inType)
```
debugging line?

deads2k commented on pkg/tools/etcd_helper.go:
```diff
@@ -101,6 +101,8 @@ func recordEtcdRequestLatency(verb, resource string, startTime time.Time) {
 type EtcdHelper struct {
 	Client EtcdGetSet
 	Codec  runtime.Codec
+	// optional, will use default conversions on api.Scheme if not set.
```
I don't see the defaulting below.

smarterclayton committed:
Use runtime.Copier instead of hardcoding api.Scheme

Allow other schemes to be supported by etcd_helper.go

runtime.Scheme.Copy() should be using the built in DeepCopy()

saad-ali commented:
Assigning @lavalamp 
Needs v1.0 triage

smarterclayton commented:
Rebased, still needs triage

lavalamp commented:
This looks fine to me-- but does it make a noticeable performance impact?

lavalamp commented:
(sorry I somehow missed this the first time around)

smarterclayton commented:
Very few people were using runtime.Scheme.Copy() - those who were were doing conversion (much less efficient than generated deep copiers) because it required marshaling to bytes.  The code problem breaks people who don't use api.Scheme with etcd (like us, because store other things in their own schemes into etcd).

lavalamp commented:
LGTM - I can't get shippable to load to see if it's a flake or not

lavalamp commented:
@quinton-hoole for second LGTM-- @davidopp is out today

lavalamp commented:
Perf improvements should be in scope.

quinton-hoole commented:
Sure... looking now.

quinton-hoole commented on pkg/tools/etcd_helper.go:
```diff
@@ -119,11 +120,13 @@ type EtcdHelper struct {
 
 // NewEtcdHelper creates a helper that works against objects that use the internal
 // Kubernetes API objects.
+// TODO: Refactor to take a runtiem.ObjectCopier
```
nit: Typo runtiem

smarterclayton replied:
How about when I refactor it :)

> On Jun 26, 2015, at 6:27 PM, Quinton Hoole <notifications@github.com> wrote:
> 
> In pkg/tools/etcd_helper.go:
> 
> > @@ -119,11 +120,13 @@ type EtcdHelper struct {
> >  
> >  // NewEtcdHelper creates a helper that works against objects that use the internal
> >  // Kubernetes API objects.
> > +// TODO: Refactor to take a runtiem.ObjectCopier
> nit: Typo runtiem
> 
> —
> Reply to this email directly or view it on GitHub.
>

quinton-hoole replied:
Agreed.  Unlikely to cause mass system instability :)

quinton-hoole commented:
LGTM, ok-to-merge.  Shippable, you're drunk, go home.
